dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java

```diff
@@ -700,47 +637,64 @@ else if ( STATIC_DIMS.contains( dimension ) )
                 return object.get();
             }
 
-            // Tracked entity attribute
+            final DimensionalObject trackedEntityDimension = getTrackedEntityDimension( dimension );
 
-            Map<String, TrackedEntityAttributeDimension> attributes = Maps.uniqueIndex( attributeDimensions,
-                TrackedEntityAttributeDimension::getUid );
-
-            if ( attributes.containsKey( dimension ) )
+            if ( trackedEntityDimension != null )
             {
-                TrackedEntityAttributeDimension tead = attributes.get( dimension );
-
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_ATTRIBUTE, null,
-                    tead.getDisplayName(), tead.getLegendSet(), null, tead.getFilter() );
+                return trackedEntityDimension;
             }
+        }
 
-            // Tracked entity data element
+        throw new IllegalArgumentException( "Not a valid dimension: " + dimension );
+    }
 
-            Map<String, TrackedEntityDataElementDimension> dataElements = Maps.uniqueIndex( dataElementDimensions,
-                TrackedEntityDataElementDimension::getUid );
+    private DimensionalObject getTrackedEntityDimension( final String dimension )
+    {
+        // Tracked entity attribute
 
-            if ( dataElements.containsKey( dimension ) )
-            {
-                TrackedEntityDataElementDimension tedd = dataElements.get( dimension );
+        final Map<String, TrackedEntityAttributeDimension> attributes = Maps.uniqueIndex( attributeDimensions,
+            TrackedEntityAttributeDimension::getUid );
 
-                return new BaseDimensionalObject( dimension, DimensionType.PROGRAM_DATA_ELEMENT, null,
-                    tedd.getDisplayName(), tedd.getLegendSet(), tedd.getProgramStage(), tedd.getFilter() );
-            }
+        if ( attributes.containsKey( dimension ) )
+        {
+            final TrackedEntityAttributeDimension tead = attributes.get( dimension );
+            final ValueType valueType = tead.getAttribute() != null ? tead.getAttribute().getValueType() : null;
```

Comment:
tead still can be null, can you please add the test to avoid null ptr exception?

Reply:
This was legacy code, but doesn't hurt adding this check just to play on the safe side :+1: 
Change applied.